File: R/boot_compare_cor.R
#' @title Comparing Correlations Between Independent Studies with Bootstrapping
#' @description
#' `r lifecycle::badge('stable')`
#'
#' A function to compare correlation coefficients between independent studies using bootstrap methods.
#' This function is intended to be used to compare the compatibility of original studies
#' with replication studies (lower p-values indicating lower compatibility).
#'
#' @param x1,y1 Numeric vectors of data values from study 1. x1 and y1 must have the same length.
#' @param x2,y2 Numeric vectors of data values from study 2. x2 and y2 must have the same length.
#' @param ... Additional arguments passed to the correlation functions.
#' @inheritParams boot_cor_test
#'
#' @details
#' This function tests for differences between correlation coefficients from independent studies
#' using bootstrap resampling methods. Unlike the `compare_cor` function, which uses Fisher's z
#' transformation or the Kraatz method with summary statistics, this function works with raw
#' data and uses bootstrapping to estimate confidence intervals and p-values.
#'
#' It is particularly useful for:
#'
#' * Comparing correlations when assumptions for parametric tests may not be met
#' * Obtaining robust confidence intervals for the difference between correlations
#' * Comparing an original study with its replication using raw data
#' * Testing if correlations from different samples are equivalent
#'
#' The function supports multiple correlation methods:
#'
#' * Standard correlation coefficients (Pearson, Kendall, Spearman)
#' * Robust correlation measures (Winsorized, percentage bend)
#'
#' The function also supports both standard hypothesis testing and equivalence/minimal effect testing:
#'
#' * For standard tests (two.sided, less, greater), the function tests whether the difference
#'   between correlations differs from the null value (typically 0).
#'
#' * For equivalence testing ("equivalence"), it determines whether the difference falls within
#'   the specified bounds, which can be set asymmetrically.
#'
#' * For minimal effect testing ("minimal.effect"), it determines whether the difference falls
#'   outside the specified bounds.
#'
#' When performing equivalence or minimal effect testing:
#' * If a single value is provided for `null`, symmetric bounds ±value will be used
#' * If two values are provided for `null`, they will be used as the lower and upper bounds
#'
#' @return A list with class "htest" containing the following components:
#'
#' * **p.value**: The p-value for the test under the null hypothesis.
#' * **parameter**: Sample sizes from each study.
#' * **conf.int**: Bootstrap confidence interval for the difference in correlations.
#' * **estimate**: Difference in correlations between studies.
#' * **stderr**: Standard error of the difference (estimated from bootstrap distribution).
#' * **null.value**: The specified hypothesized value(s) for the null hypothesis.
#' * **alternative**: Character string indicating the alternative hypothesis.
#' * **method**: Description of the correlation method used.
#' * **data.name**: Names of the input data vectors.
#' * **boot_res**: List containing the bootstrap samples for the difference and individual correlations.
#' * **call**: The matched call.
#'
#' @examples
#' # Example 1: Comparing Pearson correlations (standard test)
#' set.seed(123)
#' x1 <- rnorm(30)
#' y1 <- x1 * 0.6 + rnorm(30, 0, 0.8)
#' x2 <- rnorm(25)
#' y2 <- x2 * 0.3 + rnorm(25, 0, 0.9)
#'
#' # Two-sided test with Pearson correlation (use fewer bootstraps for example)
#' boot_compare_cor(x1, y1, x2, y2, method = "pearson",
#'                 alternative = "two.sided", R = 500)
#'
#' # Example 2: Testing for equivalence with Spearman correlation
#' # Testing if the difference in correlations is within ±0.2
#' boot_compare_cor(x1, y1, x2, y2, method = "spearman",
#'                 alternative = "equivalence", null = 0.2, R = 500)
#'
#' # Example 3: Testing with robust correlation measure
#' # Using percentage bend correlation for non-normal data
#' boot_compare_cor(x1, y1, x2, y2, method = "bendpercent",
#'                 alternative = "greater", R = 500)
#'
#' # Example 4: Using asymmetric bounds for equivalence testing
#' boot_compare_cor(x1, y1, x2, y2, method = "pearson",
#'                 alternative = "equivalence", null = c(-0.1, 0.3), R = 500)
#'
#' @name boot_compare_cor
#' @family compare studies
#' @importFrom stats median
#' @export boot_compare_cor
#'

boot_compare_cor <- function(
    x1,y1,
    x2,y2,
    alternative = c("two.sided","less", "greater",
                    "equivalence", "minimal.effect"),
    method = c("pearson", "kendall", "spearman",
               "winsorized", "bendpercent"),
    alpha = 0.05,
    null = 0,
    R = 1999,
    ...){
  DNAME <- paste(deparse(substitute(x1)), "and", deparse(substitute(y1)),
                 "vs.",
                 deparse(substitute(x2)), "and", deparse(substitute(y2)))
  nboot = R
  null.value = null
  if(!is.vector(x1) || !is.vector(x2) || !is.vector(y1) || !is.vector(y2)){
    stop("x1, x2, y1, y2  must be vectors.")
  }
  if(length(x1)!=length(y1)){
    stop("x1 and y1 do not have equal lengths.")
  }
  if(length(x2)!=length(y2)){
    stop("x2 and y2 do not have equal lengths.")
  }

  #if(TOST && null <=0){
  #  stop("positive value for null must be supplied if using TOST.")
  #}
  #if(TOST){
  #  alternative = "less"
  #}

  if(alternative %in% c("equivalence", "minimal.effect")){
    if(length(null) == 1){
      null = c(null, -1*null)
    }
    TOST = TRUE
  } else {
    if(length(null) > 1){
      stop("null can only have 1 value for non-TOST procedures")
    }
    TOST = FALSE
  }

  if(alternative != "two.sided"){
    ci = 1 - alpha*2
    intmult = c(1,1)
  } else {
    ci = 1 - alpha
    if(TOST){
      intmult = c(1,1)
    } else if(alternative == "less"){
      intmult = c(1,NA)
    } else {
      intmult = c(NA,1)
    }
  }

  alternative = match.arg(alternative)
  method = match.arg(method)

  df <- cbind(x1,y1)
  df <- df[complete.cases(df), ]
  n1 <- nrow(df)
  x1 <- df[,1]
  y1 <- df[,2]
  df <- cbind(x2,y2)
  df <- df[complete.cases(df), ]
  n2 <- nrow(df)
  x2 <- df[,1]
  y2 <- df[,2]

  if(method %in% c("bendpercent","winsorized")){
    if(method == "bendpercent"){
      r1 <- pbcor(x1, y1, ...)
      r2 <- pbcor(x2, y2, ...)
      # bootstrap
      data1 <- matrix(sample(n1, size=n1*nboot, replace=TRUE), nrow=nboot)
      bvec1 <- apply(data1, 1, .corboot_pbcor, x1, y1, ...) # A 1 by nboot matrix.
      data2 <- matrix(sample(n2, size=n2*nboot, replace=TRUE), nrow=nboot)
      bvec2 <- apply(data2, 1, .corboot_pbcor, x2, y2, ...) # A 1 by nboot matrix.
    }

    if(method == "winsorized"){
      r1 <- wincor(x1, y1, ...)
      r2 <- wincor(x2, y2, ...)

      # bootstrap
      data1 <- matrix(sample(n1, size=n1*nboot, replace=TRUE), nrow=nboot)
      bvec1 <- apply(data1, 1, .corboot_wincor, x1, y1, ...) # A 1 by nboot matrix.
      data2 <- matrix(sample(n2, size=n2*nboot, replace=TRUE), nrow=nboot)
      bvec2 <- apply(data2, 1, .corboot_wincor, x2, y2, ...) # A 1 by nboot matrix
    }


  } else {
    # correlations
    r1 <- cor(x1,y1,method = method)
    r2 <- cor(x2,y2,method = method)
    # bootstrap
    data1 <- matrix(sample(n1, size=n1*nboot, replace=TRUE), nrow=nboot)
    bvec1 <- apply(data1, 1, .corboot, x1, y1,  method = method) # A 1 by nboot matrix.
    data2 <- matrix(sample(n2, size=n2*nboot, replace=TRUE), nrow=nboot)
    bvec2 <- apply(data2, 1, .corboot, x2, y2, method = method) # A 1 by nboot matrix.
  }


  bvec <- bvec1 - bvec2
  bsort <- sort(bvec)
  boot.cint = quantile(bvec, c((1 - ci) / 2, 1 - (1 - ci) / 2))
  attr(boot.cint, "conf.level") <- ci
  # p value
  ## note method different than Efron (e.g., t-test, SMDs, etc)
  ## Derived from work of Wilcox
  if(alternative == "two.sided"){
    phat <- (sum(bvec < null.value)+.5*sum(bvec==0))/nboot
    sig <- 2 * min(phat, 1 - phat)
  }
  if(alternative == "greater"){
    sig <- 1 - sum(bvec >= null.value)/nboot
  }
  if(alternative == "less"){
    sig <- 1 - sum(bvec <= null.value)/nboot
  }
  if(alternative == "equivalence"){
    #sig2 <- 1 - sum(bvec >= -1*null.value)/nboot
    #sig = max(sig,sig2)
    sig1 = 1 - sum(bvec >= min(null.value))/nboot
    sig2 = 1 - sum(bvec <= max(null.value))/nboot
    sig = max(sig1,sig2)
  }
  if(alternative == "minimal.effect"){
    #sig2 <- 1 - sum(bvec >= -1*null.value)/nboot
    #sig = max(sig,sig2)
    sig1 = 1 - sum(bvec >= max(null.value))/nboot
    sig2 = 1 - sum(bvec <= min(null.value))/nboot
    sig = min(sig1,sig2)
  }


  if (method == "pearson") {
    # Pearson # Fisher
    method2 <- "Bootstrapped difference in Pearson's correlation"
    names(null.value) = "difference in correlation"
    rfinal = c(cor = r1-r2)
  }
  if (method == "spearman") {
    method2 <- "Bootstrapped difference in Spearman's rho"
    #  # Fieller adjusted
    rfinal = c(rho = r1-r2)
    names(null.value) = "difference in rho"
  }
  if (method == "kendall") {
    method2 <- "Bootstrapped difference in Kendall's tau"
    # # Fieller adjusted
    rfinal = c(tau = r1-r2)
    names(null.value) = "difference in tau"
  }
  if (method == "bendpercent") {
    method2 <- "Bootstrapped difference in percentage bend correlation pb"
    # # Fieller adjusted
    rfinal = c(pb = r1-r2)
    names(null.value) = "difference in pb"
  }
  if (method == "winsorized") {
    method2 <- "Bootstrapped difference in Winsorized correlation wincor"
    # # Fieller adjusted
    rfinal = c(wincor = r1-r2)
    names(null.value) = "differnce in wincor"
  }
  N = c(n1 = n1, n2 = n2)
  # Store as htest
  rval <- list(p.value = sig,
               parameter = N,
               conf.int = boot.cint,
               estimate = rfinal,
               stderr = sd(bvec,na.rm=TRUE),
               null.value = null.value,
               alternative = alternative,
               method = method2,
               data.name = DNAME,
               boot_res = list(diff = bvec,
                           r1 = bvec1,
                           r2 = bvec2),
               call = match.call())
  class(rval) <- "htest"
  return(rval)
}

File: pyTOSTER/boot_compare_cor.py
import numpy as np
from scipy import stats
import warnings

try:
    import pingouin as pg
except ImportError:
    warnings.warn("pingouin is not installed. Please install it to use robust correlation methods.")
    pg = None

def boot_compare_cor(x1, y1, x2, y2,
                     alternative='two.sided',
                     method='pearson',
                     alpha=0.05,
                     null=0,
                     R=1999):
    """
    Comparing Correlations Between Independent Studies with Bootstrapping.

    Note: Robust methods 'winsorized' and 'bendpercent' require the 'pingouin' package.
    """
    if not isinstance(x1, (list, np.ndarray)) or not isinstance(y1, (list, np.ndarray)) or \
       not isinstance(x2, (list, np.ndarray)) or not isinstance(y2, (list, np.ndarray)):
        raise TypeError("x1, y1, x2, y2 must be list or numpy array.")

    if len(x1) != len(y1) or len(x2) != len(y2):
        raise ValueError("x and y must have the same length.")

    # Remove missing values
    x1, y1 = np.array(x1), np.array(y1)
    x2, y2 = np.array(x2), np.array(y2)

    mask1 = ~np.isnan(x1) & ~np.isnan(y1)
    x1, y1 = x1[mask1], y1[mask1]

    mask2 = ~np.isnan(x2) & ~np.isnan(y2)
    x2, y2 = x2[mask2], y2[mask2]

    n1, n2 = len(x1), len(x2)

    # Bootstrap resampling
    bvec1 = np.zeros(R)
    bvec2 = np.zeros(R)

    for i in range(R):
        idx1 = np.random.choice(n1, n1, replace=True)
        idx2 = np.random.choice(n2, n2, replace=True)

        if method == 'pearson':
            bvec1[i] = stats.pearsonr(x1[idx1], y1[idx1])[0]
            bvec2[i] = stats.pearsonr(x2[idx2], y2[idx2])[0]
        elif method == 'spearman':
            bvec1[i] = stats.spearmanr(x1[idx1], y1[idx1])[0]
            bvec2[i] = stats.spearmanr(x2[idx2], y2[idx2])[0]
        elif method == 'kendall':
            bvec1[i] = stats.kendalltau(x1[idx1], y1[idx1])[0]
            bvec2[i] = stats.kendalltau(x2[idx2], y2[idx2])[0]
        elif method == 'winsorized':
            if pg is None: raise ImportError("pingouin package is required for winsorized correlation.")
            bvec1[i] = pg.corr(stats.mstats.winsorize(x1[idx1]), stats.mstats.winsorize(y1[idx1]))['r'].values[0]
            bvec2[i] = pg.corr(stats.mstats.winsorize(x2[idx2]), stats.mstats.winsorize(y2[idx2]))['r'].values[0]
        elif method == 'bendpercent':
            if pg is None: raise ImportError("pingouin package is required for percentage bend correlation.")
            bvec1[i] = pg.corr(x1[idx1], y1[idx1], method='percbend')['r'].values[0]
            bvec2[i] = pg.corr(x2[idx2], y2[idx2], method='percbend')['r'].values[0]
        else:
            raise ValueError(f"Method '{method}' not supported.")

    bvec_diff = bvec1 - bvec2

    # Calculate p-value
    if alternative == 'two.sided':
        phat = (np.sum(bvec_diff < null) + 0.5 * np.sum(bvec_diff == null)) / R
        p_value = 2 * min(phat, 1 - phat)
    elif alternative == 'greater':
        p_value = np.sum(bvec_diff <= null) / R
    elif alternative == 'less':
        p_value = np.sum(bvec_diff >= null) / R
    elif alternative == 'equivalence':
        if isinstance(null, (int, float)):
            low_eqbound, high_eqbound = -abs(null), abs(null)
        else:
            low_eqbound, high_eqbound = min(null), max(null)
        p1 = np.sum(bvec_diff >= high_eqbound) / R
        p2 = np.sum(bvec_diff <= low_eqbound) / R
        p_value = max(p1, p2)
    elif alternative == 'minimal.effect':
        if isinstance(null, (int, float)):
            low_eqbound, high_eqbound = -abs(null), abs(null)
        else:
            low_eqbound, high_eqbound = min(null), max(null)
        p1 = np.sum(bvec_diff <= high_eqbound) / R
        p2 = np.sum(bvec_diff >= low_eqbound) / R
        p_value = min(p1, p2)
    else:
        raise ValueError(f"Alternative '{alternative}' not supported.")

    # Confidence interval
    if alternative in ['equivalence', 'minimal.effect']:
        ci_level = 1 - 2 * alpha
    else:
        ci_level = 1 - alpha

    conf_int = np.quantile(bvec_diff, [(1 - ci_level) / 2, 1 - (1 - ci_level) / 2])

    # Initial correlations
    if method == 'pearson':
        r1 = stats.pearsonr(x1, y1)[0]
        r2 = stats.pearsonr(x2, y2)[0]
    elif method == 'spearman':
        r1 = stats.spearmanr(x1, y1)[0]
        r2 = stats.spearmanr(x2, y2)[0]
    elif method == 'kendall':
        r1 = stats.kendalltau(x1, y1)[0]
        r2 = stats.kendalltau(x2, y2)[0]
    elif method == 'winsorized':
        if pg is None: raise ImportError("pingouin package is required for winsorized correlation.")
        r1 = pg.corr(stats.mstats.winsorize(x1), stats.mstats.winsorize(y1))['r'].values[0]
        r2 = pg.corr(stats.mstats.winsorize(x2), stats.mstats.winsorize(y2))['r'].values[0]
    elif method == 'bendpercent':
        if pg is None: raise ImportError("pingouin package is required for percentage bend correlation.")
        r1 = pg.corr(x1, y1, method='percbend')['r'].values[0]
        r2 = pg.corr(x2, y2, method='percbend')['r'].values[0]

    estimate = r1 - r2
    stderr = np.std(bvec_diff)

    results = {
        'p.value': p_value,
        'parameter': {'n1': n1, 'n2': n2},
        'conf.int': conf_int,
        'estimate': estimate,
        'stderr': stderr,
        'null.value': null,
        'alternative': alternative,
        'method': f"Bootstrapped difference in {method} correlation",
        'boot_res': {'diff': bvec_diff, 'r1': bvec1, 'r2': bvec2}
    }
    return results
